assets/charts/control-plane/kube-apiserver/templates/kube-apiserver.yaml

```diff
@@ -189,89 +127,6 @@ spec:
-        - |
-          set -xe
-          export ADDRESS=$(cat /run/kube-apiserver/address)
-          # Make sure initContainer generated kube-apiserver address.
-          if [ -z $ADDRESS ]; then
-            echo "ADDRESS not found"
-            exit 1
-          fi
-          echo "Connecting to $ADDRESS:{{ template "port" . }}"
-          # We use TCP readiness probe and HAProxy does not reject connections if no backend is available,
-          # so we wait until kube-apiserver is available here, so readiness of haproxy container represents
-          # readiness of kube-apiserver, as kube-apiserver cannot have readiness probe set, as it listens
-          # on random IP address.
-          until nc -zv $ADDRESS {{ template "port" . }}; do sleep 1; done
-          echo "Connected"
-          # From https://github.com/docker-library/haproxy/blob/master/Dockerfile-debian.template#L70
-          exec haproxy -f /run/kube-apiserver/haproxy.cfg
-        # Readiness probe here tests that at least the apiserver was up at least once (because HAProxy waiting
-        # for apiserver to start), after that it will mean just that HAProxy is up. But this is needed for helm
-        # to perform atomic upgrades, helm can't do atomic upgrades for components without readinessProbe.
-        readinessProbe:
-          httpGet:
-            httpHeaders:
-            - name: Authorization
-              value: "{{ include "authHeader" . }}"
-            path: /healthz
-            port: {{ template "port" . }}
-            scheme: HTTPS
-      initContainers:
-      - name: config-generator
-        image: haproxy:2.1.4-alpine
-        command:
-        - /bin/sh
-        - -c
-        - |
-          set -xe
-          export ADDRESS="127.$((RANDOM%255)).$((RANDOM%255)).$(((RANDOM%254)+1))"
-          echo $ADDRESS > /run/kube-apiserver/address
-          cat <<EOF > /run/kube-apiserver/haproxy.cfg
-          defaults
-            # Do TLS passthrough
-            mode tcp
-            # Required values for both frontend and backend
-            timeout connect 5s
-            timeout client 30s
-            timeout client-fin 30s
-            timeout server 30s
-            timeout tunnel 21d
-
-          frontend kube-apiserver-internal
-            bind $POD_IP:{{ template "port" . }}
-            default_backend kube-apiserver
-
-          {{- if .Values.apiserver.exposeOnAllInterfaces }}
-          frontend kube-apiserver-external
-            bind 0.0.0.0:6443
-            default_backend kube-apiserver
-
-          {{ end }}
-          backend kube-apiserver
-            server 1 $ADDRESS:{{ template "port" . }}
-          EOF
-          cat /run/kube-apiserver/haproxy.cfg
-        volumeMounts:
-        - name: data
-          mountPath: /run/kube-apiserver
-        env:
-        - name: POD_IP
-          valueFrom:
-            fieldRef:
-              fieldPath: status.podIP
-      {{- end }}
       securityContext:
         runAsNonRoot: true
```

Comment:
To share port with bootstrap kube-apiserver, kube-apiserver must run as root, as bootstrap kube-apiserver will run as root to be able to read secret files, which we copy over SSH to the controller node.

Reply:
For the certs part there isn't a problem (HA or not) clusters wouldn't have come up. 

Also I tried multiple back and forth upgrade on k8s 1.19.1 and 1.19.2 and it seems that the root for flag is not needed either.

Reply:
>For the certs part there isn't a problem (HA or not) clusters wouldn't have come up.

What do you mean here?

>Also I tried multiple back and forth upgrade on k8s 1.19.1 and 1.19.2 and it seems that the root for flag is not needed either.

Did you verify, that bootstrap kube-apiserver and self-hosted kube-apiserver can run at the same time? This can be done e.g. using bootkube image which use --wait flag. I did and it didn't work for me otherwise, as bootstrap kube-apiserver runs as root and self-hosted one does not.

Reply:
> as bootstrap kube-apiserver will run as root to be able to read secret files

I was replying to this. I was wondering if this is the reason to run apiserver to run as root?

Reply:
About bootstrap apiserver and hosted apiserver to run simultaenously. They must have run together at some point during boot process right?

Reply:
No, currently bootstrap API server gets shut down when hosted apiserver pod is **created**. Hosted one usually crashes once or twice right now, when bootstrap is still running. The idea is to make them run simultaenously.

Reply:
>I was replying to this. I was wondering if this is the reason to run apiserver to run as root?

Yes, this is the reason to run it as root.

Reply:
> as bootstrap kube-apiserver will run as root to be able to read secret files, which we copy over SSH to the controller node.

We need to run as root to be able to read the secrets or to be able to share port? I am confused? I understand to share port we might have to run as root. But if we needed to run as root to read the secret then self-hosted kube-apiserver on current master cannot run because it is running as non-root.

Reply:
So, bootstrap kube-apiserver must run as root (and it does right now) to be able to read certificates from the host `/etc/kubernetes` directory, as we make them readable by `root` user only.

Then, to share port self-hosted kube-apiserver must use same UID as bootstrap kube-apiserver. So as bootstrap kube-apiserver runs as root, self-hosted kube-apiserver must run as root as well to be able to share the port.

Reply:
I see, gotcha.